### DISCUSSION TITLE
WithRetryPolicy(httpexpect.DontRetry) does not work with WebSockets

### DISCUSSION THREAD

**Discussion by mmatczuk** [gavv/httpexpect#182] [Q&A] [answered]
WebSockets do get retried even if `WithRetryPolicy(httpexpect.DontRetry)` is explicitly set.
Version `github.com/gavv/httpexpect/v2 v2.6.1`.

To reproduce:
- Set a deadline to WS connection
- Set WithRetryPolicy(httpexpect.DontRetry) on a request
- Run a WS test that exceeds the deadline

You can observe the errors are logged but the test happily goes on.

```
=== RUN   TestWebSocketEcho
    printer.go:116: GET /ws/echo HTTP/1.1
        Host: httpbin
        
    printer.go:133: HTTP/1.1 101 Switching Protocols 18.797083ms
        Connection: Upgrade
        Sec-Websocket-Accept: ZK3yrKgYwiXJoLiJYgp54rGPF2U=
        Upgrade: websocket
        
    printer.go:152: -> Sent: text(1)
        hello 0
        
    printer.go:171: <- Received: text(1)
        hello 0
        
    e2e_test.go:110: sent 1
    printer.go:152: -> Sent: text(1)
        hello 1
        
    printer.go:171: <- Received: text(1)
        hello 1
        
    e2e_test.go:110: sent 2
    printer.go:152: -> Sent: text(1)
        hello 2
        
    printer.go:171: <- Received: text(1)
        hello 2
        
    e2e_test.go:110: sent 3
    printer.go:152: -> Sent: text(1)
        hello 3
        
    printer.go:171: <- Received: text(1)
        hello 3
        
    e2e_test.go:110: sent 4
    printer.go:152: -> Sent: text(1)
        hello 4
        
    printer.go:171: <- Received: text(1)
        hello 4
        
    e2e_test.go:110: sent 5
    printer.go:152: -> Sent: text(1)
        hello 5
        
    printer.go:171: <- Received: close(8) AbnormalClosure(1006)
        unexpected EOF
        
    assertion.go:239: 
        expected: message types are equal
        
        assertion:
          Request("GET").Expect().Websocket().Expect().TextMessage()
        
        expected value:
          text(1)
        
        actual value:
          close(8)
    e2e_test.go:110: sent 6
    printer.go:152: -> Sent: text(1)
        hello 6
        
    assertion.go:239: 
        failed to write to websocket
          write tcp 172.26.0.5:39060->172.26.0.4:3128: write: broken
          pipe
        
        assertion:
          Request("GET").Expect().Websocket().WriteText().WriteMessage()
    e2e_test.go:110: sent 7
    e2e_test.go:110: sent 8
```

**Comment by gavv** [comment]
Hi, thanks for report.

Could you please show the test that reproduces it? And if possible could you show which operation in the source code of the test triggered deadline and was retried?

**Comment by mmatczuk** [comment]
Sure the test can be found [here](https://github.com/saucelabs/forwarder/blob/25ff36a4661c77566316f3595822ad38ebc5fe17/e2e/e2e_test.go#L92). The server code is [here](https://github.com/saucelabs/forwarder/blob/25ff36a4661c77566316f3595822ad38ebc5fe17/httpbin/ws.go#L10). 

What's unusual it that the traffic goes through a proxy and the proxy sets a 5s t/o (deadline now+5s). 
The same can be obtained by running http.Server with ReadTimeout and WriteTimeout.

**Comment by gavv** [comment]
Thanks.

Sorry for stupid question, but could you point me where exactly do you observe a retry?

From what I see you're using [testing.T](https://github.com/saucelabs/forwarder/blob/25ff36a4661c77566316f3595822ad38ebc5fe17/e2e/framework.go#L96) as Reporter, so failures are non-fatal (httpexpect will use t.Errorf), i.e. after failure test will continue, but will be marked as failed. And you continue sending messages in a loop disregarding of failures. After WriteText() fails in message 6, websocket marks itself as failed, and further calls to WriteText() from loop become no-op.

Am I missing something?

BTW if you need fatal failures, you can use httpexpect.NewRequireReporter.

**Comment by mmatczuk** [comment, marked as answer]
Thanks for the detailed description. I double checked and indeed there is no retry. I was fooled by "websocket marks itself as failed, and further calls to WriteText() from loop become no-op".

**Comment by gavv** [comment]
Converted this to discussion.

> Thanks for the detailed description. I double checked and indeed there is no retry. I was fooled by "websocket marks itself as failed, and further calls to WriteText() from loop become no-op".

Yep, all httpexpect methods work like this. If reporter does not interrupt test, after failure subsequent assertion on same or child objects become no-op.
